fe/fe-core/src/main/java/com/starrocks/lake/compaction/CompactionScheduler.java

```diff
@@ -0,0 +1,412 @@
+import com.starrocks.lake.LakeTablet;
+import com.starrocks.lake.Utils;
+import com.starrocks.lake.proto.CompactRequest;
+import com.starrocks.lake.proto.CompactResponse;
+import com.starrocks.rpc.BrpcProxy;
+import com.starrocks.rpc.LakeService;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.service.FrontendOptions;
+import com.starrocks.system.Backend;
+import com.starrocks.system.SystemInfoService;
+import com.starrocks.transaction.BeginTransactionException;
+import com.starrocks.transaction.GlobalTransactionMgr;
+import com.starrocks.transaction.TabletCommitInfo;
+import com.starrocks.transaction.TransactionState;
+import com.starrocks.transaction.VisibleStateWaiter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import javax.validation.constraints.NotNull;
+
+public class CompactionScheduler extends Daemon {
+    private static final Logger LOG = LogManager.getLogger(CompactionScheduler.class);
+    private static final String HOST_NAME = FrontendOptions.getLocalHostAddress();
+    private static final long LOOP_INTERVAL_MS = 500L;
+    private static final long TXN_TIMEOUT_SECOND = 86400L;
+    private static final long MIN_COMPACTION_INTERVAL_MS_ON_SUCCESS = 3000L;
+    private static final long MIN_COMPACTION_INTERVAL_MS_ON_FAILURE = 6000L;
+    private static final long PARTITION_CLEAN_INTERVAL_SECOND = 30;
+
+    private boolean finishedWaiting = false;
+    private long waitTxnId = -1;
+    private final CompactionManager compactionManager;
+    private final SystemInfoService systemInfoService;
+    private final GlobalTransactionMgr transactionMgr;
+    private final GlobalStateMgr stateMgr;
+    private final Map<PartitionIdentifier, CompactionContext> runningCompactions;
+    private long lastPartitionCleanTime;
+
+    CompactionScheduler(@NotNull CompactionManager compactionManager, @NotNull SystemInfoService systemInfoService,
+                        @NotNull GlobalTransactionMgr transactionMgr, @NotNull GlobalStateMgr stateMgr) {
+        super("COMPACTION_DISPATCH", LOOP_INTERVAL_MS);
+        this.compactionManager = compactionManager;
+        this.systemInfoService = systemInfoService;
+        this.transactionMgr = transactionMgr;
+        this.stateMgr = stateMgr;
+        this.runningCompactions = Maps.newHashMap();
+        this.lastPartitionCleanTime = System.currentTimeMillis();
+    }
+
+    @Override
+    protected void runOneCycle() {
+        cleanPartition();
+
+        // Schedule compaction tasks only when this is a leader FE and all edit logs have finished replay.
+        if (!stateMgr.isReady()) {
+            return;
+        }
+        if (allCommittedTransactionsBeforeRestartHaveFinished()) {
```

Comment:
Please add comments for why add this check.

Reply:
updated